### DISCUSSION TITLE
Permanent redirects in app directory

### DISCUSSION THREAD

**Discussion by danjebs** [vercel/next.js#46871] [Help]
### Summary

I am trying to set up permanent redirects in the the app directory in a similar manner to how they are implemented in `getServerSideProps()`. I have looked into using `redirect()` - but I don't see any options to do permanent redirects (`redirect()` issues 307s).

The redirect logic is complex enough that redirects have to be dynamically generated on the fly, so it's not possible to use `redirects` in `next.config.js`.

Currently it is handled in `getServerSideProps` like so:
```
export const getServerSideProps = async () => {
  const data = await fetch('/api/get-stuff')

  const { redirect, ...rest } = data

  if (redirect) {
    return {
      redirect: {
        permanent: true,
        destination: redirect,
      }
    }
  }

  return {
    props: rest
  }
}
```

I know it is possible to issue redirects in middleware. But that would require either making a separate API call in middleware to generate just the redirect, or re-running the `getData()` API call in middleware as well as in `page.jsx` (and this won't be de-duped). So there comes a performance overhead in this approach.

Is this something fundamentally incompatible with app directory and server components? Is it somewhere on the roadmap? Or maybe I am missing something.

**Comment by MahmoudBakr22** [comment]
I have the exact same problem but in this issue https://github.com/vercel/next.js/issues/49148 

@timneutkens mentioned that the only way to do dynamic permanent redirect is using middleware which is a huge drawback as in Middleware I have no idea if I am gonna need to redirect this url or not
